File: JanperCab/5-JanperCab/ClientApp/src/app/_models/duraform-option/DuraformOptionTypeDto.ts
import { DuraformOptionAngledShelfDto } from './DuraformOptionAngledShelfDto';
import { DuraformOptionRollerShutterFrameDto } from './DuraformOptionRollerShutterFrameDto';
import { DuraformOptionDto } from 'src/app/_models/duraform-option/DuraformOptionDto';
import { DuraformOptionPaneFrameDto } from './DuraformOptionPaneFrameDto';
import { DuraformOptionFoldBackDto } from './DuraformOptionFoldBackDto';
import { DuraformOptionNoFaceDto } from './DuraformOptionNoFaceDto';
import { DuraformOptionTypeKey } from '../../_enums/DuraformOptionTypeKey';
import { DuraformOptionDoubleSidedDto } from './DuraformOptionDoubleSidedDto';
import { DuraformOptionMicrowaveFrameDto } from './DuraformOptionMicrowaveFrameDto';

export class DuraformOptionTypeDto {
  id: DuraformOptionTypeKey;
  name: string;
  type: string;

  static GetDuraformOptionInstance(
    optionType: DuraformOptionTypeDto,
    optionValues: any
  ): DuraformOptionDto {
    let option;

    switch (optionType.id) {
      case DuraformOptionTypeKey.NoFaceRoute:
        option = new DuraformOptionNoFaceDto();
        option.$type = optionType.type;
        break;
      case DuraformOptionTypeKey.DoubleSided:
        option = new DuraformOptionDoubleSidedDto();
        option.$type = optionType.type;
        option.hasProfile = optionValues.hasProfile;
        break;
      case DuraformOptionTypeKey.FoldBack:
        option = new DuraformOptionFoldBackDto();
        option.$type = optionType.type;
        option.hasProfile = optionValues.hasProfile;
        option.thickness = optionValues.thickness;
        option.foldingType = optionValues.foldingType;
        option.leftLength = optionValues.leftLength;
        option.rightLength = optionValues.rightLength;
        break;
      case DuraformOptionTypeKey.PaneFrame:
        option = new DuraformOptionPaneFrameDto();
        option.$type = optionType.type;
        option.columns = optionValues.columns;
        option.rows = optionValues.rows;
        break;
      case DuraformOptionTypeKey.RollerShutter:
        option = new DuraformOptionRollerShutterFrameDto();
        option.$type = optionType.type;
        option.topSize = optionValues.topSize;
        option.leftSize = optionValues.leftSize;
        option.rightSize = optionValues.rightSize;
        break;
      case DuraformOptionTypeKey.MicrowaveFrame:
        option = new DuraformOptionMicrowaveFrameDto();
        option.$type = optionType.type;
        option.topSize = optionValues.topSize;
        option.bottomSize = optionValues.bottomSize;
        option.leftSize = optionValues.leftSize;
        option.rightSize = optionValues.rightSize;
        break;
      case DuraformOptionTypeKey.AngledShelf:
        option = new DuraformOptionAngledShelfDto();
        option.$type = optionType.type;
        option.sideOne = optionValues.sideOne;
        option.sideTwo = optionValues.sideTwo;
        option.isDoubleSided = optionValues.isDoubleSided;
        break;
      default:
        throw new Error('Unsupported Option Type');
    }

    option.duraformOptionTypeId = optionType.id;
    return option;
  }
}

File: JanperCab/3-Application/Dtos/DuraformOptionType/DuraformOptionTypeDto.cs
using _1_Domain.Enum;
using _3_Application.Dtos.DuraformOption;
using System;

namespace _3_Application.Dtos.DuraformOptionType
{
    public class DuraformOptionTypeDto
    {
        public DuraformOptionTypeKey Id { get; set; }

        public string Name { get; set; }

        public string Type
        {
            get
            {
                switch (Id)
                {
                    case DuraformOptionTypeKey.NoFaceRoute:
                        return typeof(DuraformOptionNoFaceDto).AssemblyQualifiedName;
                    case DuraformOptionTypeKey.DoubleSided:
                        return typeof(DuraformOptionDoubleSidedDto).AssemblyQualifiedName;
                    case DuraformOptionTypeKey.FoldBack:
                        return typeof(DuraformOptionFoldBackDto).AssemblyQualifiedName;
                    case DuraformOptionTypeKey.PaneFrame:
                        return typeof(DuraformOptionPaneFrameDto).AssemblyQualifiedName;
                    case DuraformOptionTypeKey.RollerShutter:
                        return typeof(DuraformOptionRollerShutterFrameDto).AssemblyQualifiedName;
                    case DuraformOptionTypeKey.MicrowaveFrame:
                        return typeof(DuraformOptionMicrowaveFrameDto).AssemblyQualifiedName;
                    case DuraformOptionTypeKey.AngledShelf:
                        return typeof(DuraformOptionAngledShelfDto).AssemblyQualifiedName;
                    default:
                        throw new NotImplementedException("Duraform Option Type Not Recognized");
                }
            }
        }
    }
}